SpatialGDK/Source/SpatialGDK/Private/SpatialView/EntityComponentRecord.cpp

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "SpatialView/EntityComponentRecord.h"
+
+namespace SpatialGDK
+{
+
+void EntityComponentRecord::AddComponent(Worker_EntityId EntityId, ComponentData Data)
+{
+	const EntityComponentId Id = { EntityId, Data.GetComponentId() };
+
+	// If the component is recorded as removed then transition to complete-updated.
+	// otherwise record it as added.
+	if (ComponentsRemoved.RemoveSingleSwap(Id))
+	{
+		UpdateRecord.AddComponentDataAsUpdate(EntityId, MoveTemp(Data));
+	}
+	else
+	{
+		ComponentsAdded.Push(EntityComponentData{ EntityId, MoveTemp(Data) });
+	}
+}
+
+void EntityComponentRecord::RemoveComponent(Worker_EntityId EntityId, Worker_ComponentId ComponentId)
+{
+	const EntityComponentId Id = { EntityId, ComponentId };
+	const EntityComponentData* FoundComponentAdded = ComponentsAdded.FindByPredicate(EntityComponentIdEquality{ Id });
```

Comment:
How big can these arrays be ? That's a lot of linear searches.

Reply:
We need to take more data on this but I suspect normally small. I don't think doing things non-linearly is faster until the list gets huge (even a binary search is probably not better as it involves keeping the data sorted; but it was what originally happened).
There is an alternative formulation of all of this that doesn't do all the linear searches that we might use but it is more complicated and makes slightly different assumptions. This is the easiest formation to read and understand.

Reply:
Worth noting this is really just formalising what the GDK already does (or should do if bug free).

Reply:
I do not know what is the association between ops and these objects, but the fact that it can contain several entities makes me wonder if it could contain the entire visible state of the world on connection, which would make the initial checkout unreasonably slow, but subsequent ones faster.

Reply:
It would but no worse than it already does. An on initial checkout things tend not to be so bad because you only really have stuff added, so the checking the remove queue is basically always a no-op.
But there are a couple of ways to get around this. This particular view delta has been made such that you can create it in a background thread if that becomes a problem. You can also add a maximum number of ops to process in a tick.
The other formulations of view-deltas don't have the linear search problem (specifically the faster one sorts the data then iterates over it all once) but limiting the number of ops would still make it faster if we need to amortise over several ticks.

Reply:
This is the reason I made the Unity GDK version of this always run in a background thread though. But it had worse problems as it needed to serialise things too and the implementation I did back then was worse too.